ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,819 @@
+			let signed = SignedTransaction::new(public_tx.with_signature(signature, chain_id))?;
+			let client = self.client.read().clone().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+			let client = client.upgrade().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+			match client.miner().import_own_transaction(&*client as &MiningBlockChainClient, signed.into()) {
+				Ok(_) => trace!("Public transaction added to queue"),
+				Err(err) => trace!("Failed to add transaction to queue, error: {:?}", err),
+			}
+			//Remove from store for signing
+			match self.transactions_for_signing.lock().remove(&private_hash) {
+				Ok(_) => {}
+				Err(err) => trace!("Failed to remove transaction from signing store, error: {:?}", err),
+			}
+		} else {
+			//Add signature to the store
+			match self.transactions_for_signing.lock().add_signature(&private_hash, tx.signature()) {
+				Ok(_) => trace!("Signature stored for private transaction"),
+				Err(err) => trace!("Failed to add signature to signing store, error: {:?}", err),
+			}
+		}
+		Ok(())
+	}
+
+	fn last_required_signature(&self, desc: &PrivateTransactionSigningDesc, sign: Signature) -> Result<bool, EthcoreError>  {
+		let existing_signatures = desc.received_signatures.clone();
+		if existing_signatures.contains(&sign) {
+			return Ok(false);
+		}
+		let state_hash = keccak(desc.state.clone());
+		match recover(&sign, &state_hash) {
+			Ok(public) => {
+				let sender = public_to_address(&public);
+				let validators = desc.validators.clone();
+				match validators.contains(&sender) {
+					true => {
+						Ok(existing_signatures.len() + 1 == validators.len())
+					}
+					false => {
+						trace!("Sender's state doesn't correspond to validator's");
+						Err(PrivateTransactionError::StateIncorrect.into())
+					}
+				}
+			}
+			Err(err) => {
+				trace!("Sender's state doesn't correspond to validator's, error {:?}", err);
+				Err(err.into())
+			}
+		}
+	}
+
+	/// Broadcast the private transaction message to the chain
+	fn broadcast_private_transaction(&self, message: Bytes) {
+		self.notify(|notify| notify.broadcast(ChainMessageType::PrivateTransaction, message.clone()));
+	}
+
+	/// Broadcast signed private transaction message to the chain
+	fn broadcast_signed_private_transaction(&self, message: Bytes) {
+		self.notify(|notify| notify.broadcast(ChainMessageType::SignedPrivateTransaction, message.clone()));
+	}
+
+	fn encrypt(&self, _contract_address: &Address, data: &[u8]) -> Result<Bytes, EthcoreError> {
+		//TODO: retrieve key from secret store using contract
+		let init_key: Bytes = "cac6c205eb06c8308d65156ff6c862c62b000b8ead121a4455a8ddeff7248128d895692136f240d5d1614dc7cc4147b1bd584bd617e30560bb872064d09ea325".from_hex().unwrap();
+		let key: Bytes = init_key[..INIT_VEC_LEN].into();
+
+		let iv = initialization_vector();
+		let mut encrypted = Vec::with_capacity(data.len() + iv.len());
+		encrypted.extend(repeat(0).take(data.len()));
```

Comment:
Having iterator here seems excessive, maybe just do `vec![0; data.len()+iv.len()]`, and then use `split_at_mut` to get that buffer for encryption?